001-clarifications-eval.txt

```diff
@@ -0,0 +1,754 @@
+    described in 2.5.4.4.3 Memory Location Description Operations are
+    performed if valid. Otherwise, the DWARF expression is ill-formed.
+
+    If the evaluation of a DWARF expression encounters an evaluation
+    error, then the result is an evaluation error.
+
+    If a DWARF expression is ill-formed, then the result is undefined.
+
+    The following sections detail the rules for when a DWARF expression
+    is ill-formed or results in an evaluation error.
+    --------------------------------------------------------------------
+
+Insert the following subsection under 2.5 (renumbering subsequent
+subsections):
+
+    --------------------------------------------------------------------
+    2.5.1 DWARF Expression Evaluation Context
+
+    A DWARF expression is evaluated in a context that can include a
+    number of context elements. If multiple context elements are
+    specified then they must be self consistent or the result of the
+    evaluation is undefined. The context elements that can be specified
+    are:
+
+    1. A current result kind
+
+        The kind of result required by the DWARF expression evaluation.
+        If specified it can be a location description or a value.
+
+    2. A current thread
+
+        The target architecture thread identifier of the source program
+        thread of execution for which a user presented expression is
+        currently being evaluated.
+
+        It is required for operations that are related to target
+        architecture threads.
+
+        [non-normative] For example: the DW_OP_regval_type operation.
+
+    3. A current call frame
+
+        The target architecture call frame identifier. It identifies a
+        call frame that corresponds to an active invocation of a
+        subprogram in the current thread. It is identified by its
+        address on the call stack. The address is referred to as the
+        Canonical Frame Address (CFA). The call frame information is
+        used to determine the CFA for the call frames of the current
+        thread’s call stack (see 6.4 Call Frame Information).
+
+        It is required for operations that specify target architecture
+        registers to support virtual unwinding of the call stack.
+
+        [non-normative] For example: the DW_OP_*reg* operations.
+
+        If specified, it must be an active call frame in the current
+        thread. Otherwise the result is undefined.
+
+        If it is the currently executing call frame, then it is termed
+        the top call frame.
+
+    4. A current program location
+
+        The target architecture program location corresponding to the
+        current call frame of the current thread.
+
+        The program location of the top call frame is the target
+        architecture program counter for the current thread. The call
+        frame information is used to obtain the value of the return
+        address register to determine the program location of the other
```

Comment:
It sounds like the target has to have a return address *register*. But in general a return address can be located at a place other than a certain register.

Reply:
CFI works in terms of target architecture registers. These registers may be hardware registers, or pseudo registers. Every target has a PC register. As CFI is used to virtually unwind the call stack it is determining what the register values would be if it were about to execute the call. So the return address is the PC register value when the stack is virtually unwound.

It is true that CFI provides the location description of each register as it virtually unwinds, and that location may not be a physical register due to spilling or clobbering. However, in a DWARF operation expression, the DW_OP_reg et al operations are "as if" the call stack was virtually unwound. An implementation can use CFI to get the location description of where the register is actually present (or that it is not available) and can use that to access the value of the register.

The text in the CFI part which @ccoutant has not added yet does describe this more completely than the DWARF 5 spec does.

Reply:
In "6.4.1 Structure of Call Frame Information" it defines:

> return_address_register (unsigned LEB128)
> An unsigned LEB128 constant that indicates which column in the rule table represents the return address of the function. Note that this column might not correspond to an actual machine register.

Section "6.4.4 Call Frame Calling Address" also mentions the Return Address register. Perhaps the use here should be similarly capitalized, or a reference to section 6.4.4 added here to clarify what return address register means?

Reply:
> Section "6.4.4 Call Frame Calling Address" also mentions the Return Address register. Perhaps the use here should be similarly capitalized, or a reference to section 6.4.4 added here to clarify what return address register means?

I think a reference would help. Capitalization or putting the phrase inside quotations would additionally reduce the chances of confusion. But how would the committee react to renaming "Return Address Register" to just "Return Address". It sounds like the "register" part is a historical remnant.